In [159]:
#importing and pre-processing data
import pandas as pd
import numpy as np
import re
import seaborn as sns
import matplotlib.pyplot as plt

from sklearn.preprocessing import LabelEncoder
from sklearn.preprocessing import MinMaxScaler
from sklearn.preprocessing import StandardScaler,OneHotEncoder

from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report 

from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
import tensorflow as tf


In [160]:
# Read the data into a Pandas DataFrame
loan_info_df = pd.read_csv('Resources/loan_training_dataset.csv')

loan_info_df.head()


INCOME  AGE  EXPERIENCE MARITAL_STATUS HOUSE_OWNERSHIP CAR_OWNERSHIP  \
0  1303834   23           3         single          rented            no   
1  7574516   40          10         single          rented            no   
2  3991815   66           4        married          rented            no   
3  6256451   41           2         single          rented           yes   
4  5768871   47          11         single          rented            no   

            PROFESSION           STATE  CURRENT_JOB_YRS  RISK_FLAG  
0  Mechanical Engineer  Madhya Pradesh                3          0  
1   Software Developer     Maharashtra                9          0  
2     Technical Writer          Kerala                4          0  
3   Software Developer          Odisha                2          1  
4        Civil Servant      Tamil Nadu                3          1

In [161]:
loan_info_df.describe(include='all').T

count unique            top    freq            mean  \
INCOME           252000.0    NaN            NaN     NaN  4997116.665325   
AGE              252000.0    NaN            NaN     NaN       49.954071   
EXPERIENCE       252000.0    NaN            NaN     NaN       10.084437   
MARITAL_STATUS     252000      2         single  226272             NaN   
HOUSE_OWNERSHIP    252000      3         rented  231898             NaN   
CAR_OWNERSHIP      252000      2             no  176000             NaN   
PROFESSION         252000     51      Physician    5957             NaN   
STATE              252000     28  Uttar Pradesh   29143             NaN   
CURRENT_JOB_YRS  252000.0    NaN            NaN     NaN        6.333877   
RISK_FLAG        252000.0    NaN            NaN     NaN           0.123   

                            std      min        25%        50%        75%  \
INCOME           2878311.013611  10310.0  2503015.0  5000694.5  7477502.0   
AGE                   17.063855     21.0       35.0       50.0       65.0   
EXPERIENCE              6.00259      0.0        5.0       10.0       15.0   
MARITAL_STATUS              NaN      NaN        NaN        NaN        NaN   
HOUSE_OWNERSHIP             NaN      NaN        NaN        NaN        NaN   
CAR_OWNERSHIP               NaN      NaN        NaN        NaN        NaN   
PROFESSION                  NaN      NaN        NaN        NaN        NaN   
STATE                       NaN      NaN        NaN        NaN        NaN   
CURRENT_JOB_YRS        3.647053      0.0        3.0        6.0        9.0   
RISK_FLAG              0.328438      0.0        0.0        0.0        0.0   

                       max  
INCOME           9999938.0  
AGE                   79.0  
EXPERIENCE            20.0  
MARITAL_STATUS         NaN  
HOUSE_OWNERSHIP        NaN  
CAR_OWNERSHIP          NaN  
PROFESSION             NaN  
STATE                  NaN  
CURRENT_JOB_YRS       14.0  
RISK_FLAG              1.0

In [162]:
loan_info_df.describe().T

count          mean           std      min        25%  \
INCOME           252000.0  4.997117e+06  2.878311e+06  10310.0  2503015.0   
AGE              252000.0  4.995407e+01  1.706385e+01     21.0       35.0   
EXPERIENCE       252000.0  1.008444e+01  6.002590e+00      0.0        5.0   
CURRENT_JOB_YRS  252000.0  6.333877e+00  3.647053e+00      0.0        3.0   
RISK_FLAG        252000.0  1.230000e-01  3.284379e-01      0.0        0.0   

                       50%        75%        max  
INCOME           5000694.5  7477502.0  9999938.0  
AGE                   50.0       65.0       79.0  
EXPERIENCE            10.0       15.0       20.0  
CURRENT_JOB_YRS        6.0        9.0       14.0  
RISK_FLAG              0.0        0.0        1.0

### Step 2: Create the labels set (`y`)  from the “Risk_Flag” column, and then create the features (`X`) DataFrame from the remaining columns.

In [163]:
# Separate the data into labels and features
# Separate the y variable, the labels
y = loan_info_df['RISK_FLAG']

# Separate the X variable, the features
X = loan_info_df.drop(columns=['RISK_FLAG'])

In [164]:
# Review the y variable Series
y.head()

0    0
1    0
2    0
3    1
4    1
Name: RISK_FLAG, dtype: int64

In [165]:
# Review the X variable DataFrame
X.head()

INCOME  AGE  EXPERIENCE MARITAL_STATUS HOUSE_OWNERSHIP CAR_OWNERSHIP  \
0  1303834   23           3         single          rented            no   
1  7574516   40          10         single          rented            no   
2  3991815   66           4        married          rented            no   
3  6256451   41           2         single          rented           yes   
4  5768871   47          11         single          rented            no   

            PROFESSION           STATE  CURRENT_JOB_YRS  
0  Mechanical Engineer  Madhya Pradesh                3  
1   Software Developer     Maharashtra                9  
2     Technical Writer          Kerala                4  
3   Software Developer          Odisha                2  
4        Civil Servant      Tamil Nadu                3

### 3. Encode the categorical variables from the features data using `label_encoder` and `get_dummies`.

In [166]:
loan_info_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 252000 entries, 0 to 251999
Data columns (total 10 columns):
 #   Column           Non-Null Count   Dtype 
---  ------           --------------   ----- 
 0   INCOME           252000 non-null  int64 
 1   AGE              252000 non-null  int64 
 2   EXPERIENCE       252000 non-null  int64 
 3   MARITAL_STATUS   252000 non-null  object
 4   HOUSE_OWNERSHIP  252000 non-null  object
 5   CAR_OWNERSHIP    252000 non-null  object
 6   PROFESSION       252000 non-null  object
 7   STATE            252000 non-null  object
 8   CURRENT_JOB_YRS  252000 non-null  int64 
 9   RISK_FLAG        252000 non-null  int64 
dtypes: int64(5), object(5)
memory usage: 19.2+ MB


In [167]:
# Encode Ownership columns with label encoder
# Initialize LabelEncoder

label_encoder = LabelEncoder()
loan_info_df['CAR_OWNERSHIP'] = label_encoder.fit_transform(loan_info_df['CAR_OWNERSHIP'])
loan_info_df['HOUSE_OWNERSHIP'] = label_encoder.fit_transform(loan_info_df['HOUSE_OWNERSHIP'])
loan_info_df['MARITAL_STATUS'] = label_encoder.fit_transform(loan_info_df['MARITAL_STATUS'])
loan_info_df.head()

INCOME  AGE  EXPERIENCE  MARITAL_STATUS  HOUSE_OWNERSHIP  CAR_OWNERSHIP  \
0  1303834   23           3               1                2              0   
1  7574516   40          10               1                2              0   
2  3991815   66           4               0                2              0   
3  6256451   41           2               1                2              1   
4  5768871   47          11               1                2              0   

            PROFESSION           STATE  CURRENT_JOB_YRS  RISK_FLAG  
0  Mechanical Engineer  Madhya Pradesh                3          0  
1   Software Developer     Maharashtra                9          0  
2     Technical Writer          Kerala                4          0  
3   Software Developer          Odisha                2          1  
4        Civil Servant      Tamil Nadu                3          1

In [168]:
# Apply get_dummies to the 'Profession' and 'State' columns
loan_info_df = pd.get_dummies(loan_info_df, columns=['PROFESSION', 'STATE'], dtype=int)
loan_info_df.head()

INCOME  AGE  EXPERIENCE  MARITAL_STATUS  HOUSE_OWNERSHIP  CAR_OWNERSHIP  \
0  1303834   23           3               1                2              0   
1  7574516   40          10               1                2              0   
2  3991815   66           4               0                2              0   
3  6256451   41           2               1                2              1   
4  5768871   47          11               1                2              0   

   CURRENT_JOB_YRS  RISK_FLAG  PROFESSION_Air Traffic Controller  \
0                3          0                                  0   
1                9          0                                  0   
2                4          0                                  0   
3                2          1                                  0   
4                3          1                                  0   

   PROFESSION_Analyst  ...  STATE_Puducherry  STATE_Punjab  STATE_Rajasthan  \
0                   0  ...                 0             0                0   
1                   0  ...                 0             0                0   
2                   0  ...                 0             0                0   
3                   0  ...                 0             0                0   
4                   0  ...                 0             0                0   

   STATE_Sikkim  STATE_Tamil Nadu  STATE_Telangana  STATE_Tripura  \
0             0                 0                0              0   
1             0                 0                0              0   
2             0                 0                0              0   
3             0                 0                0              0   
4             0                 1                0              0   

   STATE_Uttar Pradesh  STATE_Uttarakhand  STATE_West Bengal  
0                    0                  0                  0  
1                    0                  0                  0  
2                    0                  0                  0  
3                    0                  0                  0  
4                    0                  0                  0  

[5 rows x 87 columns]

### Step 4: Split the data into training and testing datasets by using `train_test_split`.

In [171]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42)

### 5. Scale the data using `MinMaxScaler`

In [172]:
# Create a MinMaxScaler instances
scaler = MinMaxScaler()

# Fit the StandardScaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

ValueError: could not convert string to float: 'single'

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [173]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
logistic_regression_model = LogisticRegression(solver='lbfgs', random_state=42)

# Fit the model using training data
lr_model = logistic_regression_model.fit(X_train, y_train)

ValueError: could not convert string to float: 'single'

### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [174]:
# Make a prediction using the testing data
prediction_test = logistic_regression_model.predict(X_test)
pd.DataFrame({'Predictions': prediction_test, 'Actual': y_test})

ValueError: could not convert string to float: 'single'

### Step 3: Evaluate the model’s performance by doing the following:

* Generate a confusion matrix.

* Print the classification report.

In [175]:
# Generate a confusion matrix for the model
confusion_matrix(y_test, prediction_test)

NameError: name 'prediction_test' is not defined

In [176]:
# Print the classification report for the model
target_names = ['Non Risky Application', 'Flagged Risky Application']
print(classification_report(y_test, prediction_test, target_names=target_names))

NameError: name 'prediction_test' is not defined

### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (non-risk loan) and `1` (high-risk loan) labels?

**Answer:** The logistic regression model was 88% accurate at predicting the healthy vs high-risk loan labels

In [ ]:
# Define the model - deep neural net
number_input_features = X_train.shape[1] 
hidden_nodes_layer1 =  8
hidden_nodes_layer2 = 5

nn = tf.keras.models.Sequential()

# First hidden layer
nn.add(
    tf.keras.layers.Dense(units=hidden_nodes_layer1, input_dim=number_input_features, activation="relu")
)

# Second hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer2, activation="relu"))

# Output layer
nn.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

# Check the structure of the model
nn.summary()

/Users/irynavitkovych/anaconda3/envs/dev/lib/python3.10/site-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense (Dense)                   │ (None, 8)              │           728 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 5)              │            45 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_2 (Dense)                 │ (None, 1)              │             6 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 779 (3.04 KB)

 Trainable params: 779 (3.04 KB)

 Non-trainable params: 0 (0.00 B)

In [ ]:
# Compile the model
nn.compile(loss = "binary_crossentropy", optimizer = "adam", metrics=["accuracy"])

In [ ]:
# Train the model
fit_model = nn.fit(X_train_scaled,y_train,epochs=100)

Epoch 1/100
5907/5907 ━━━━━━━━━━━━━━━━━━━━ 2s 250us/step - accuracy: 0.8713 - loss: 0.3911
Epoch 2/100
5907/5907 ━━━━━━━━━━━━━━━━━━━━ 2s 253us/step - accuracy: 0.8775 - loss: 0.3574
Epoch 3/100
5907/5907 ━━━━━━━━━━━━━━━━━━━━ 1s 247us/step - accuracy: 0.8764 - loss: 0.3554
Epoch 4/100
5907/5907 ━━━━━━━━━━━━━━━━━━━━ 1s 247us/step - accuracy: 0.8779 - loss: 0.3494
Epoch 5/100
5907/5907 ━━━━━━━━━━━━━━━━━━━━ 2s 261us/step - accuracy: 0.8786 - loss: 0.3443
Epoch 6/100
5907/5907 ━━━━━━━━━━━━━━━━━━━━ 1s 248us/step - accuracy: 0.8796 - loss: 0.3388
Epoch 7/100
5907/5907 ━━━━━━━━━━━━━━━━━━━━ 1s 251us/step - accuracy: 0.8772 - loss: 0.3421
Epoch 8/100
5907/5907 ━━━━━━━━━━━━━━━━━━━━ 1s 249us/step - accuracy: 0.8772 - loss: 0.3403
Epoch 9/100
5907/5907 ━━━━━━━━━━━━━━━━━━━━ 1s 249us/step - accuracy: 0.8779 - loss: 0.3385
Epoch 10/100
5907/5907 ━━━━━━━━━━━━━━━━━━━━ 2s 258us/step - accuracy: 0.8778 - loss: 0.3387
Epoch 11/100
5907/5907 ━━━━━━━━━━━━━━━━━━━━ 1s 249us/step - accuracy: 0.8783 - loss: 0.33

In [ ]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

1969/1969 - 0s - 200us/step - accuracy: 0.8768 - loss: 0.3409
Loss: 0.3408711850643158, Accuracy: 0.876793622970581


# Result Explanation

# Loss: 0.3409
The loss value represents how well or poorly the model's predictions match the true values in the test data. A lower loss indicates better performance.
The loss value of 0.3409 suggests that, on average, the model's predictions are fairly close to the actual risk flags in the test data, but lower result preffered.

# Accuracy: 0.8768
Accuracy is the proportion of correct predictions out of the total number of predictions made. An accuracy of 0.8829 means that the model correctly predicted the risk flag for approximately 88.29% of the test samples.
In the context of loan risk prediction, an accuracy of 87.68% is generally considered quite good, indicating that the model is correctly identifying risky and non-risky applicants most of the time. However, it's important to consider the balance of the dataset and other metrics like precision, recall, and F1-score to ensure that the model is not biased towards the majority class.